CONTRIBUTING.md

```diff
@@ -150,6 +152,185 @@ meson test -C build [tests...]
+ * immediately before the function definition.
+ */
+/* MTF_MOCK */
+struct foo *foo_create(void);
+
+/* MTF_MOCK */
+void
+foo_destroy(struct foo *foo);
+
+/* This function is not mockable because there is no MTF_MOCK comment. */
+void foo_bar(void);
+
+/* Include generated file that defines macros that wrap mockable functions. */
+#if HSE_MOCKING
+#include "foo_ut.h"
+#endif
+
+#endif
+```
+
+###### `foo_print.h`
+
+```c
+/* SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ * SPDX-FileCopyrightText: Copyright YYYY Micron Technology, Inc.
+ */
+
+#ifndef FOO_PRINT_H
+#define FOO_PRINT_H
+
+/* MTF_MOCK_DECL(foo_print) */
+
+/* MTF_MOCK */
+void foo_print(void);
+
+#if HSE_MOCKING
+#include "foo_print_ut.h"
+#endif
+
+#endif
+```
+
+###### `foo.c`
+
+```c
+/* SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ * SPDX-FileCopyrightText: Copyright YYYY Micron Technology, Inc.
+ */
+
+/* Declare that this file implements functions in the `foo` and `foo_print`
+ * groups. These are typically defined at the top of the file immediately after
+ * the copyright notice.
+ */
+#define MTF_MOCK_IMPL_foo
+#define MTF_MOCK_IMPL_foo_print
+
+/* Header files containing prototypes of mockable functions MUST be included
+ * after the MTF_MOCK_IMPL definitions.
+ */
+#include "abc.h"
+#include "bar.h"
+#include "foo.h"
+#include "foo_print.h"
+#include "xyz.h"
+
+/* This function is mockable by virtue of the MTF_MOCK annotation in foo.h. No
+ * annoation is needed here.
+ */
+struct foo *
+foo_create()
+{
+    // ...
+}
+
+void foo_destroy(struct foo *foo)
+{
+    // ...
+}
+
+void foo_print(void)
+{
+    // ...
+}
+
+/* This function is not mockable since the prototype in foo.h is not annotated
+ * with MTF_MOCK.
+ */
+void foo_bar(void)
+{
+    // ...
+}
+
+/* Must be at end of source file. Shoud have one include for each MTF_MOCK_IMPL
+ * definition at the top of the source file. These generated files define hooks
+ * to access the mocked and real versions of mockable functions.
+ */
+#if HSE_MOCKING
+#include "foo_ut_impl.i"
+#include "foo_print_ut_impl.i"
+#endif
+```
+
+##### Troubleshooting
+
+When implementing mocking, you may run into one of the following issues:
+
+###### `mapi_idx_foo_create undeclared`
+
+Build does not know that `foo.h` contains mock declarations.
+
+Add `foo.h` to the list of mocked headers in
+[`tests/mocks/meson.build`](./tests/mocks/meson.build).
+
+###### `unknown type name mtfm_foo_create_fp`
+
+Indicates the following code is missing from the bottom of `foo.h`:
+
+```c
+#ifdef HSE_MOCKING
+#include "foo_ut.h"
+#endif
+```
+
+###### `multiple definition of mtfm_foo_foo_create_getreal`
+
+More than on source file has `#include "cn_foo_ut_impl.i"`.
+
+###### `undefined reference to mtfm_foo_foo_create_get`
+
+Could indicate one of two things:
+
+* The source file with `foo_create()` is not compiled
+* The source file with `foo_create()` is missing `#include "cn_foo_ut_impl.i"`
```

Comment:
I think there are some other ways to manifest this that require rebuilding from a clean slate to fix it (I suspect due to something broken regarding dependencies).

Reply:
I remember seeing this in the past, but not anymore. If you do run into it, let me know since it might be a Meson issue :).